### PR TITLE
Check email with hashed value

### LINKED ISSUE #13: Allow checking against an email, without providing the email itself
We need a way to check if an email exists in our database, without providing the actual email. This helps in not sharing the email, as this "breaks" the confidentiality agreements that might exist for a company.

In order to do this, we'll need to use a 1-way hashing function, which will match every email to a consistent hash, that cannot be reversed. As a result:

* If a user provides a hash that exists in our database, we can say with certainty that the hash is blacklisted
* If a user provides a hash that does not exist in our database, we will not be aware of the actual email address of the non-blacklisted user

In order to do this, we need to:

* Include a new field in the email table, called `hashed_value`, which will contain the hashed value of the email
* Include a different method, that will check against the hashed value

This new value should:

* always be computed from the `lowercase` value of the email, so that it is consistent
* be computed on every save
* be read-only in the admin interface

### PR DESCRIPTION
**Pull request by nzisouli**
Allow checking against an email, without providing the email itself but a hash. Closes #13

### CODE REVIEW

**Comment by nzisouli** [comment]
Finish work

File: bouncer/blacklist/admin.py [hunk 1]
```diff
@@ -2,7 +2,14 @@
 from blacklist import models
 
 
-@admin.register(models.IPEntry, models.EmailEntry, models.EmailHostEntry)
+@admin.register(models.IPEntry, models.EmailHostEntry)
 class EntryAdmin(admin.ModelAdmin):
-    fields = ["entry_value", "reason"]
-    search_fields = ["entry_value"]
+    fields = ["entry_value", "lower_case_entry_value", "reason"]
+    search_fields = ["lower_case_entry_value", "entry_value"]
```
Comment (akalipetis) [inline bouncer/blacklist/admin.py]: There is no need for `lower_case_entry_value` to be searchable, `entry_value` will suffice.

File: bouncer/blacklist/admin.py [hunk 2]
```diff
@@ -2,7 +2,14 @@
...
+
+
+@admin.register(models.EmailEntry)
+class EmailEntryAdmin(admin.ModelAdmin):
+    fields = ["entry_value", "lower_case_entry_value", "hashed_value", "reason"]
+    search_fields = ["lower_case_entry_value", "entry_value"]
```
Comment (akalipetis) [inline bouncer/blacklist/admin.py]: There is no need for `lower_case_entry_value` to be searchable, `entry_value` will suffice. Also, it would be nice to be able to search against the `hashed_value` here, so that admins can also search against that.

File: bouncer/blacklist/signals.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+@receiver(pre_save, sender=models.EmailEntry)
+def email_entry_handler(sender, instance, **kwargs):
+    """
+    Handler that assigns to lower_case_entry_value the entry_value.lower()
+    """
+
+    instance.lower_case_entry_value = instance.entry_value.lower()
+    hashed_email = sha256(instance.lower_case_entry_value.encode())
```
Comment (akalipetis) [inline bouncer/blacklist/signals.py]: `email_hasher` would be a better name for this variable.

**Review by akalipetis — CHANGES_REQUESTED** [review]
Nice one! Please include a test, that will verify that an email entry will have get a known SHA256 hash after the email is saved.

Please make sure that this email has a different case (ie uppercase), to make sure the whole process (lower -> hash) is followed.

File: bouncer/blacklist/admin.py [hunk 4]
```diff
@@ -2,7 +2,14 @@
...
+@admin.register(models.IPEntry, models.EmailHostEntry)
 class EntryAdmin(admin.ModelAdmin):
-    fields = ["entry_value", "reason"]
+    fields = ["entry_value", "lower_case_entry_value", "reason"]
     search_fields = ["entry_value"]
+
+
+@admin.register(models.EmailEntry)
+class EmailEntryAdmin(admin.ModelAdmin):
+    fields = ["entry_value", "lower_case_entry_value", "hashed_value", "reason"]
```
Comment (akalipetis) [inline bouncer/blacklist/admin.py]: Please completely remove the `lower_case_entry_value` from the admin, as it gives no value being here.

File: bouncer/blacklist/brain.py [hunk 5]
```diff
@@ -30,3 +30,13 @@ def is_email_host_blacklisted(host):
...
         return True
     return False
+
+
+def is_email_hash_blacklisted(email_hash):
+    """
+    Checks if email hashed value is in EmailEntry model
+    """
+
+    if models.EmailEntry.objects.filter(hashed_value=email_hash):
```
Comment (akalipetis) [inline bouncer/blacklist/brain.py]: What if the hash is given as uppercase and the one you save in the database is lowercase?

File: bouncer/blacklist/signals.py [hunk 6]
```diff
@@ -19,6 +20,8 @@ def email_entry_handler(sender, instance, **kwargs):
     """
 
     instance.lower_case_entry_value = instance.entry_value.lower()
+    email_hasher = sha256(instance.lower_case_entry_value.encode())
+    instance.hashed_value = email_hasher.hexdigest()
```
Comment (akalipetis) [inline bouncer/blacklist/signals.py]: We should convert this to either `upper()`, or `lower()`, to make sure we know which case to check against.

File: bouncer/blacklist/tests.py [hunk 7]
```diff
@@ -139,6 +141,22 @@ def test_is_email_host_blacklisted_with_not_blacklisted_host(self):
 
         self.assertFalse(is_email_host_blacklisted(self.not_blacklisted_host))
 
+    def test_is_email_hash_blacklisted_with_blacklisted_hash(self):
+        """
+        Test that checking a blacklisted email hash correctly returns True.
+        """
+
+        hashed_email = sha256(self.lower_case_blacklisted_email.encode())
+        self.assertTrue(is_email_hash_blacklisted(hashed_email.hexdigest()))
```
Comment (akalipetis) [inline bouncer/blacklist/tests.py:150]: Add another check for the hash being in `.upper()`.

**Review by akalipetis — CHANGES_REQUESTED** [review]
Nice one, just another edge case and LGTM 🚀

**Review by akalipetis — APPROVED** [review]
Nice one, LGTM 🐻
